src/VisualStudio/Core/Def/Implementation/ProjectSystem/AnalyzerConfigDocumentAsSolutionItemHandler.cs

```diff
@@ -0,0 +1,136 @@
+
+#nullable enable
+
+using System;
+using System.Collections.Generic;
+using System.Composition;
+using System.Linq;
+using System.Threading;
+using EnvDTE;
+using EnvDTE80;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CodeFixes;
+using Microsoft.CodeAnalysis.Editor.Shared.Utilities;
+using Microsoft.CodeAnalysis.Extensions;
+using Microsoft.CodeAnalysis.Host.Mef;
+using Microsoft.CodeAnalysis.Options;
+using Microsoft.VisualStudio.Shell;
+using Roslyn.Utilities;
+using Task = System.Threading.Tasks.Task;
+
+namespace Microsoft.VisualStudio.LanguageServices.Implementation.ProjectSystem
+{
+    [Export, Shared]
+    internal partial class AnalyzerConfigDocumentAsSolutionItemHandler : IDisposable
+    {
+        private static readonly string LocalRegistryPath = $@"Roslyn\Internal\{nameof(AnalyzerConfigDocumentAsSolutionItemHandler)}\";
+        private static readonly Option<bool> NeverShowAgain = new Option<bool>(nameof(AnalyzerConfigDocumentAsSolutionItemHandler), nameof(NeverShowAgain),
+            defaultValue: false, storageLocations: new LocalUserProfileStorageLocation(LocalRegistryPath + nameof(NeverShowAgain)));
+
+        private readonly VisualStudioWorkspace _workspace;
+        private readonly IThreadingContext _threadingContext;
+
+        private DTE? _dte;
+
+        [ImportingConstructor]
+        [Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
+        public AnalyzerConfigDocumentAsSolutionItemHandler(
+            SVsServiceProvider serviceProvider,
+            VisualStudioWorkspace workspace,
+            IThreadingContext threadingContext)
+        {
+            _workspace = workspace;
+            _threadingContext = threadingContext;
+
+            _workspace.WorkspaceChanged += OnWorkspaceChanged;
+        }
+
+        public void Initialize(IServiceProvider serviceProvider)
+        {
+            _dte = (DTE)serviceProvider.GetService(typeof(DTE));
+        }
+
+        void IDisposable.Dispose()
+        {
+            _workspace.WorkspaceChanged -= OnWorkspaceChanged;
+        }
+
+        private void OnWorkspaceChanged(object sender, WorkspaceChangeEventArgs e)
+        {
+            // Check if a new analyzer config document was added and we have a non-null DTE instance.
+            if (e.Kind != WorkspaceChangeKind.AnalyzerConfigDocumentAdded ||
+                _dte == null)
+            {
+                return;
+            }
+
+            // Check if added analyzer config document is at the root of the current solution.
+            var analyzerConfigDocumentFilePath = e.NewSolution.GetAnalyzerConfigDocument(e.DocumentId)?.FilePath;
+            var analyzerConfigDirectory = PathUtilities.GetDirectoryName(analyzerConfigDocumentFilePath);
+            var solutionDirectory = PathUtilities.GetDirectoryName(e.NewSolution?.FilePath);
```

Comment:
Why can we trust `e.NewSolution` to be non-null when calling `GetAnalyzerConfigDocument` and not when getting the `FilePath`? Is this an extension method quirk?

Reply:
Hmm that is an oversight, I will fix it here.